pkg/operation/botanist/dns.go

```diff
@@ -225,6 +270,31 @@ func (b *Botanist) GenerateTerraformDNSConfig(name, hostedZoneID, targetType str
 	}
 }
 
+func (b *Botanist) generateTerraformAlicloudDNSConfig(domain, targetType string, value string, purposeInternalDomain bool) (map[string]interface{}, error) {
+	secret, err := b.getDomainCredentials(purposeInternalDomain)
+	if err != nil {
+		return nil, err
+	}
+
+	name := secret.GetAnnotations()[common.DNSDomain]
+	if name == "" {
+		return nil, fmt.Errorf("Can't find annotaion %s in Secret %s/%s", common.DNSDomain, secret.Name, secret.Namespace)
+	}
+
+	host := strings.TrimSuffix(domain, "."+name)
+	if host == "" || host == domain {
+		return nil, fmt.Errorf("Domain %s is not a valid domain", domain)
+	}
```

Comment:
Why do we have the above validation steps here?

The ShootDNSHostedZone admission plugin (maybe rename to ShootDNS now) should have it: https://github.com/gardener/gardener/blob/master/plugin/pkg/shoot/dnshostedzone/admission.go

Reply:
It seems that we don't check secret for internal domain in admission.go.  This is why I put the check in dns.go just like other providers. Maybe it is not the best timing to change too much code in alicloud dns provider enablement. Can I just let it be currently?

Reply:
I took another look. The checks for the internal and default domain annotations are done here:
https://github.com/gardener/gardener/blob/master/pkg/operation/garden/garden.go#L68-L93
Please add what is needed for Alicloud there and ignore my previous comment.

Reply:
This is not the right place to put domain name check.
This function only checks secrets. It has no context of expected domain, which is dynamic like Shoot API Server Domain.

Can we just leave it unchanged? There are some exception checks like https://github.com/gardener/gardener/blob/master/pkg/operation/botanist/dns.go#L156, which should be covered by https://github.com/gardener/gardener/blob/master/plugin/pkg/shoot/dnshostedzone/admission.go. 
Maybe we need to do enhancement in another PR. I agree with you that precondition check should be covered by addmission controller. So that we will find issues at the first place and avoid to check logs after error happens.

Reply:
The Shoot domain is already validated in the static validation (pkg/apis/garden/validation).
This is checking the domain annotation on the default domain or internal domain secret. Why do you say it's not the right place?

The current place is not correct. We can't leave it unchanged, sorry. The check whether the domain annotation can be done when reading, and when it fails, the secret is not even considered later.

Reply:
OK, I understand your concern.  This is a duplicated check. I also need to do some enhancement in next commit. Please check it then.

Here is my understanding domain related secrets and it is not quite related to the previous conversation. Correct me if I have some mis-understood.
1. **Internal Domain**(Only used for Shoot API Server?)
As internal domain secret has a provider and credentials, it could be verified statically. I find the verification code in https://github.com/gardener/gardener/blob/master/pkg/operation/botanist/dns.go#L255 when shoot is being reconcied.  
Further enhancement: Can we add credential check in https://github.com/gardener/gardener/blob/master/plugin/pkg/shoot/dnshostedzone/admission.go . Because internal domain secret is a precondition for shoot cluster. Fail fast might be a better way.
2. **Default Domain**
Credential check is also in https://github.com/gardener/gardener/blob/master/pkg/operation/botanist/dns.go#L255. 
Add this check in https://github.com/gardener/gardener/blob/master/plugin/pkg/shoot/dnshostedzone/admission.go . ?
3. **logic to get secret related to shoot domain is not consistent** https://github.com/gardener/gardener/blob/master/pkg/operation/botanist/dns.go#L235 VS.
https://github.com/gardener/gardener/blob/master/plugin/pkg/shoot/dnshostedzone/admission.go#L160-L191

Reply:
Thank you!

1. This secret is read once during startup of GCM and it could be verified there. However, there are multiple cases where the credentials for setting up DNS might come from: Domain Secret, Cloud Provider Secret, explicitly referenced secret (`.spec.dns.secretName`). The function here https://github.com/gardener/gardener/blob/master/pkg/operation/botanist/dns.go#L255 is getting the respective secret (which depends on the configuration). The ShootDNSHostedZone admission plugin is completely unrelated with the internal domain secret. No validation there.
1. Same as for internal domain, the default domains are read only during start-up. Validation could go there. However, in both cases it's difficult as we would have to validate based on the configured provider (route53, clouddns, ...). Can't we simply continue as it is and not put any more cloud specific logic in places where it doesn't need to be necessarily? With #370 the DNS controllers get separated and it will become much easier to validate autonomously. Take a look at #215 and what was needed to integrate openstack-designate. I see that alicloud-dns is little special due to non-existing hosted zones, however, apart from that the mentioned PR is a good "reference implementation".
3. Not sure what you mean, you didn't provide any further information. Can you elaborate? Anyway, I guess it boils down to what I wrote before: There are multiple cases where the credentials might come from.

Reply:
For 3rd item:

The **sequences** of Secrets fetching and checking is not the same.

In admission controller, only ONE secret is checked. The sequence is **Default -> .spec.dns.secretName ->  Cloud Provider Secret.**

BTW, for Default secret, there is no credential check while the latters are checked. Why default secret is skipped. 

However, in dns.go, the sequence is **.spec.dns.secretName->Default->Cloud Provider Secret**.

Reply:
OK, thank's for the explanation and agreed, the order should be the same. Maybe you can refactor these parts and create a dedicated function that tells the caller which secret is to be used (default/referenced secret/cloud provider secret), call it in admission plugin and botanist.

Reply:
Hi @rfranzke , 
I still have questions regarding the useage of default domain secret.
I think this credential should be hold by gardener landscape. 

When we try to add a Shoot's external domain DNS record, shouldn't we use credentials provided by this shoot (Gardener Tenant)? 

Maybe this is a business question regarding who pays money. I will not change the current code and will post an Issue, if it is not clarified before. How do you suggest?

Reply:
OK, Thanks for your kind long explanations and I'm now more clear about the secrets used by domains. Will follow up some code soon.

Reply:
Cool. Sorry for not having proper documentation in the first place...